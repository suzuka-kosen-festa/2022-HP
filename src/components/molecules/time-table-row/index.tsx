import type { FC } from "react";
import { useMemo } from "react";
import dayjs from "dayjs";
import tw, { css } from "twin.macro";
import type { TimeTableRowProperties } from "./types/model";
import { TimeTableElement } from "../../atoms/time-table-element";
import { LiveEvent } from "../../../api/prisma";

const TimeTableRowBox = tw.ul`grid h-time-table-row gap-0.5 md:(h-time-table-row-md gap-3)`;

const TimeTableRow: FC<TimeTableRowProperties> = ({ events, intervals, stage }) => {
  const templateRows = useMemo(() => intervals.map(interval => `${interval}fr`).join(" "), [intervals]);
  const data = useMemo(() => {
    const arr: Array<LiveEvent | "gap"> = [];
    events.forEach((event, i, a) => {
      if (i === 0) {
        return dayjs(event.start_time).format("HH:mm") === "09:00" ? arr.push(event) : arr.push("gap", event);
      }
      if (i === a.length - 1) {
        return dayjs(event.end_time).format("HH:mm") === "16:00" ? arr.push(event) : arr.push("gap", event);
      }
      return dayjs(a[i - 1].end_time).format("HH:mm") === dayjs(a[i].start_time).format("HH:mm")
        ? arr.push(event)
        : arr.push("gap", event);
    });
    return arr;
  }, [events]);
  // eslint-disable-next-line consistent-return
  const color = useMemo(() => {
    switch (stage) {
      case "main":
        return tw`border-primary-red bg-red-50 border md:border-5`;
      case "sub":
        return tw`border-primary-blue bg-blue-50 border md:border-5`;
      case "live":
        return tw`border-primary-purple bg-purple-50 border md:border-5`;
      case "game":
        return tw`border-primary-green bg-green-50 border md:border-5`;
      // no default
    }
  }, [stage]);
  return (
    <TimeTableRowBox
      css={[
        css`
          grid-template-rows: ${templateRows};
        `,
        color,
      ]}
    >
      {data.map((event, i) =>
        // eslint-disable-next-line react/no-array-index-key
        event === "gap" ? <div key={i} /> : <TimeTableElement event={event} key={i} />,
      )}
    </TimeTableRowBox>
  );
};

export { TimeTableRow };
