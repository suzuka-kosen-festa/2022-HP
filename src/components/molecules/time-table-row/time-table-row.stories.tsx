import type { ComponentStoryObj, ComponentMeta } from "@storybook/react";
import { ComponentPropsWithoutRef } from "react";
import { TimeTableRow } from ".";

type T = typeof TimeTableRow;
type Story = ComponentStoryObj<T>;
type Meta = ComponentMeta<T>;

const args: ComponentPropsWithoutRef<T> = {
  events: [
    {
      id: 1,
      title: "オープニング",
      descriptions: "空白の2年間、たくさんの協力を得て遂に開催！溜め込んだものを全て出し切れ！高専祭開幕です！",
      date: "2022-10-30 09:00",
      venue: "メインステージ",
      start_time: "2022-10-30 09:00",
      end_time: "2022-10-30 09:30",
      stage: "main",
    },
    {
      id: 2,
      title: "音楽部コンサート",
      descriptions: "音楽部による、メインステージ最初を飾るセレモニー！盛り上がっていきましょー！",
      date: "2022-10-30 09:30",
      venue: "メインステージ",
      start_time: "2022-10-30 09:30",
      end_time: "2022-10-30 10:00",
      stage: "main",
    },
    {
      id: 3,
      title: "常識力バトル",
      descriptions: "高専生の常識をクイズ形式で競うバトル！高専生としての常識がある学生は誰か！？",
      date: "2022-10-30 10:00",
      venue: "メインステージ",
      start_time: "2022-10-30 10:00",
      end_time: "2022-10-30 10:30",
      stage: "main",
    },
    {
      id: 4,
      title: "勝山組",
      descriptions: "今年もよさこい勝山組の参加！鈴フェス大賞のよさこいを見逃すな！",
      date: "2022-10-30 10:45",
      venue: "メインステージ",
      start_time: "2022-10-30 10:45",
      end_time: "2022-10-30 11:15",
      stage: "main",
    },
    {
      id: 5,
      title: "へぇ王",
      descriptions: "先生達が自分の研究を紹介！4年生は卒検部を考えるいい機会に！高専生、中学生に向けた勉強イベント！",
      date: "2022-10-30 11:30",
      venue: "メインステージ",
      start_time: "2022-10-30 11:30",
      end_time: "2022-10-30 12:00",
      stage: "main",
    },
    {
      id: 6,
      title: "腕1GP",
      descriptions: "高専祭で腕相撲の1番を決める！一般参加ありの白熱バトルイベント！",
      date: "2022-10-30 12:30",
      venue: "メインステージ",
      start_time: "2022-10-30 12:30",
      end_time: "2022-10-30 13:00",
      stage: "main",
    },
    {
      id: 7,
      title: "TTTGP",
      descriptions:
        "その名もTop of The TikTok GP！TikTokで流行ったダンス、音源、Tik Tokerなど TikTokに関するクイズを出題！",
      date: "2022-10-30 13:15",
      venue: "メインステージ",
      start_time: "2022-10-30 13:15",
      end_time: "2022-10-30 13:45",
      stage: "main",
    },
    {
      id: 8,
      title: "EXBALLON",
      descriptions: "仲間を信じて、風船割って、四字熟語をより早く当てるチーム競技！！",
      date: "2022-10-30 14:00",
      venue: "メインステージ",
      start_time: "2022-10-30 14:00",
      end_time: "2022-10-30 14:45",
      stage: "main",
    },
  ],
  intervals: [2, 2, 2, 1, 2, 1, 2, 2, 2, 1, 2, 1, 3, 5],
  stage: "main",
};

export default {
  args,
  argTypes: {
    events: {
      description: "Events data",
      control: {
        type: "array",
      },
    },
    intervals: {
      description: "Intervals of element",
      control: {
        type: "array",
      },
    },
    stage: {
      description: "Live stage",
      control: {
        type: "select",
        options: ["main", "sub", "live", "game"],
      },
    },
  },
  component: TimeTableRow,
} as Meta;

export const Default: Story = {};
