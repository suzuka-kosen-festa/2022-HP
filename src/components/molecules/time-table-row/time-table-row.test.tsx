import { composeStories } from "@storybook/testing-react";
import "@testing-library/jest-dom";
import { render } from "@testing-library/react";
import * as stories from "./time-table-row.stories";

const { Default } = composeStories(stories);

describe("(components) molecules/time-table-row", () => {
  beforeAll(() => {
    const portalRoot = document.createElement("div");
    portalRoot.setAttribute("id", "modal");
    document.body.appendChild(portalRoot);
  });
  test("to be molecules", () => {
    const { container } = render(<Default />);
    expect(container).toBeMolecule();
  });
  test("take snap shot", () => {
    const { container } = render(<Default />);
    expect(container).toMatchSnapshot();
  });
});
