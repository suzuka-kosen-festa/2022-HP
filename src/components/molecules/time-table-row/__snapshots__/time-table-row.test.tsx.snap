// Vitest Snapshot v1

exports[`(components) molecules/time-table-row > take snap shot 1`] = `
<div>
  <ul
    class="css-1vf1wns-TimeTableRow"
  >
    <li
      class="css-1ycdo6t-TimeTableElement"
    >
      <time
        class="css-1phduix-TimeTableElement"
      >
        <span
          class="css-zu2x60-TimeTableElement"
        >
          09:00
        </span>
        09:30
      </time>
      <p
        class="css-1phduix-TimeTableElement"
      >
        オープニング
      </p>
    </li>
    <li
      class="css-1ycdo6t-TimeTableElement"
    >
      <time
        class="css-1phduix-TimeTableElement"
      >
        <span
          class="css-zu2x60-TimeTableElement"
        >
          09:30
        </span>
        10:00
      </time>
      <p
        class="css-1phduix-TimeTableElement"
      >
        音楽部コンサート
      </p>
    </li>
    <li
      class="css-1ycdo6t-TimeTableElement"
    >
      <time
        class="css-1phduix-TimeTableElement"
      >
        <span
          class="css-zu2x60-TimeTableElement"
        >
          10:00
        </span>
        10:30
      </time>
      <p
        class="css-1phduix-TimeTableElement"
      >
        常識力バトル
      </p>
    </li>
    <div />
    <li
      class="css-1ycdo6t-TimeTableElement"
    >
      <time
        class="css-1phduix-TimeTableElement"
      >
        <span
          class="css-zu2x60-TimeTableElement"
        >
          10:45
        </span>
        11:15
      </time>
      <p
        class="css-1phduix-TimeTableElement"
      >
        勝山組
      </p>
    </li>
    <div />
    <li
      class="css-1ycdo6t-TimeTableElement"
    >
      <time
        class="css-1phduix-TimeTableElement"
      >
        <span
          class="css-zu2x60-TimeTableElement"
        >
          11:30
        </span>
        12:00
      </time>
      <p
        class="css-1phduix-TimeTableElement"
      >
        へぇ王
      </p>
    </li>
    <div />
    <li
      class="css-1ycdo6t-TimeTableElement"
    >
      <time
        class="css-1phduix-TimeTableElement"
      >
        <span
          class="css-zu2x60-TimeTableElement"
        >
          12:30
        </span>
        13:00
      </time>
      <p
        class="css-1phduix-TimeTableElement"
      >
        腕1GP
      </p>
    </li>
    <div />
    <li
      class="css-1ycdo6t-TimeTableElement"
    >
      <time
        class="css-1phduix-TimeTableElement"
      >
        <span
          class="css-zu2x60-TimeTableElement"
        >
          13:15
        </span>
        13:45
      </time>
      <p
        class="css-1phduix-TimeTableElement"
      >
        TTTGP
      </p>
    </li>
    <div />
    <li
      class="css-1ycdo6t-TimeTableElement"
    >
      <time
        class="css-1phduix-TimeTableElement"
      >
        <span
          class="css-zu2x60-TimeTableElement"
        >
          14:00
        </span>
        14:45
      </time>
      <p
        class="css-1phduix-TimeTableElement"
      >
        EXBALLON
      </p>
    </li>
  </ul>
</div>
`;
